### DISCUSSION TITLE
Time delta 구현 상황 및 이슈

### DISCUSSION THREAD

**Discussion by ddoyoon** [bcaitech1/p4-dkt-lastffang#37] [EDA]
## 구현 방법
- 원래 데이터의 `Timestamp` 는 해당 문제를 풀기 **시작한** 시점입니다. (제가 어제 peer session 에서 끝난 시점으로 잘못 설명했습니다 😭)
- 동일한 user, testId (시험지) 에서 Timestamp 기준으로 sorting 했을 때, 인접한 두 문제 중 뒤쪽 문제를 풀기 시작한 시점이 앞쪽 문제를 푸는게 끝난 시점으로 여기고, 그 차이를 초 단위로 계산했습니다.
  - 이 값을 time delta 라고 부르고 있습니다. 
- 이렇게 하면 마지막 문제는 푸는데 소요된 시간 계산이 안 되므로, 뒤에서 두번째 문제와 동일한 시간을 부여했습니다. 
  - 평균값 등 통계량으로 대체할 수 있으나, 한 시험지 내에서도 문제의 난이도/소요시간이 대체로 점점 증가한다고 가정했습니다.
  - 이때 이 두 문제의 KnowledgeTag 가 다르거나, 학생이 문제를 순서대로 풀지 않는 경우도 있어 적절한 근사가 아닐 수도 있을 듯 합니다.

## 발견한 이슈
아래의 내용은 [여기서](https://github.com/bcaitech1/p4-dkt-lastffang/blob/timedelta/notebooks/timedelta-discussion.ipynb) 예시와 함께 확인하실 수 있습니다!
- Time delta 가 0 이 되는 경우
  - 제 방법으로는 한 시험지에 문제가 한개일 때를 커버하지 못하는데, 데이터에 그런 경우는 없는 것 같습니다.
  - **인접한 두 Timestamp 가 완전히 동일한 경우가 있습니다.** 두 문제를 풀기 시작한 시점이 동일하다는 것인데, 그럼 한 문제에 여러 문제가 딸려있는 것으로 해석하면 좋을지 명확하지 않습니다.
 - Time delta 가 (의미 없이) 매우 큰 값으로 계산되는 경우
   - 동일한 시험지의 문항들을 꼭 연속적으로 풀지는 않을 수도 있습니다.
   - 학생이 동일한 시험지를 여러 번 푸는 경우도 있는데, 현재의 구현체로는 동일 user/시험지가 여러번 연달아 나오면 무조건 time delta 를 계산하게 되어있습니다. 이 경우는 threshold 를 주는 방법으로 해결할 수 있을 것 같긴 합니다..!

**Comment by newspring97** [comment]
혹시 Time delta가 0이 되는 경우에는 인접한 두 timestamp가 동일한 경우 다음의 문제 풀이 시작 시간이 time delta가 0이 아닌 경우에 비해 유의미하게 크기 차이가 있었나요?
만약 일반적인 경우의 time delta에 비해 time delta가 0이 되는 경우의 다음 문제와의 time delta가 크게 책정된다면 여러 문제가 딸려있는 것으로 해석할 수 있을 것 같습니다.

**Comment by GY-Jeong** [comment]
time stamp를 적용할 때 time delta 값을 그대로 넣는것 이외에 다른 feature가 하나 생각나서 적어봅니다!

수업시간에 time delta 분석 내용 중, 22초 근방(정확한 수치는 다시 봐봐야합니다)에서 정답률이 가장 높게 나왔습니다.

그러면 22초에서 +- 일정 범위에 푼 내용을 1로, 나머지를 0으로 주거나 하는 방식은 어떨까요????

**Reply by gonipark** [reply]
고고고곡

**Comment by ddoyoon** [comment]
저녁에 보내드린 csv 파일들에서 결측치를 채운 방법 공유해드립니다!
현재는 제 주관이 들어간 부분도 많고 다양한 통계량으로 결측치를 대체할 수 있어서, 아이디어가 있으시다면 반영해서 새로 만들어보겠습니다.

우선은 user, testID 로 groupby 하고 인접한 문항의 Timestamp 차이로 각 문제의 풀이 시간을 구했습니다. 
이때 풀이 시간이 구해지지 않는 각 testID 의 마지막 문항과 15분이 넘어가는 시간은 0으로 초기화하고 이후에 다른 통계량으로 채웠습니다.
(제가 피어세션에서 각 시험지의 마지막 문항은 항상 시간이 구해지지 않는다고 했는데, 같은 시험지라도 사용자마다 문제를 푸는 순서가 다를 수 있어서 항상 성립하는 말은 아닙니다 👀)

결측치는 아래 순서대로 채웠는데, 같은 조건이더라도 answerCode 까지 동일한 경우가 없다면 answerCode 만 달리했을 때의 평균 시간으로 대체했습니다.
1. 다른 유저들이 같은 문항을 풀어서 같은 answerCode 낸 평균 시간
2. 같은 유저가 같은 유형의 다른 문제들을 풀어서 같은 answerCode 낸 평균 시간
3. 다른 유저들이 같은 유형을 풀어서 같은 answerCode 낸 평균 시간

test dataset 의 경우 1~3 으로도 채워지지 않는 값이 있어서 같은 유저가 모든 문항에서 같은 answerCode 를 냈을 때 걸린 평균 시간으로 일단 채웠고,
answerCode 가 -1 인 경우는 answerCode 가 0, 1 일 때의 평균 소요 시간을 구해 다시 평균을 냈습니다.

코드도 지저분한데 말로 설명하려니까 더 지저분하네요 😂 이해되지 않는 부분이 있다면 언제든 알려주세요.

일단 마음에 걸리는건 제 구현의 편의상 (?) answerCode 가 다를 때의 평균 시간이 높은 우선순위를 가졌다는건데 (예를 들어 1번 조건+answerCode 다를 때가 2번조건보다 우선적으로 채워집니다), 실제로 여기에 해당하는 케이스가 많은지는 아직 확인해보지 않았습니다..!
